pyro/poutine/messenger.py

```diff
@@ -22,11 +40,8 @@ def __init__(self):
         pass
 
     def __call__(self, fn):
-        def _wraps(*args, **kwargs):
-            with self:
-                return fn(*args, **kwargs)
-        _wraps.msngr = self
-        return _wraps
+        wraps = _bound_partial(partial(_context_wrap, self, fn))
```

Comment:
Why not simply `_bound_partial(_context_wrap, self, fn)`?

Reply:
For this to work, I will need to change `bound_partial` to 

```
return partial(partial(self.func, *self.args, **self.keywords), instance)
```

from

```
return partial(self.func, instance)
```

The reason is that instance should be bound in that case not to `self.func` but rather to the `fn` argument of `_context_wrap`. Maybe there is a simpler way to do this that I am missing out. Let me know if you want me to make this change.

Reply:
ok, thanks for explaining, this is subtler than I thought!